File: smash/solver/derived_type/mwd_options.f90
!%      (MWD) Module Wrapped and Differentiated.
!%
!%      Type
!%      ----
!%
!%      - OptionsDT
!%
!%          ======================== =======================================
!%          `Variables`              Description
!%          ======================== =======================================
!%          ``comm``                 Common_OptionsDT
!%          ======================== =======================================
!%
!%      Subroutine
!%      ----------
!%
!%      - OptionsDT_initialise
!%      - OptionsDT_copy

module mwd_options

    use mwd_setup !% only: SetupDT
    use mwd_common_options !% only: Common_OptionsDT, Common_OptionsDT_initialise
    use mwd_optimize_options !% only: Optimize_OptionsDT, Optimize_OptionsDT_initialise

    implicit none

    type OptionsDT

        type(Common_OptionsDT) :: comm
        type(Optimize_OptionsDT) :: optimize

    end type OptionsDT

contains

    subroutine OptionsDT_initialise(this, setup)

        implicit none

        type(OptionsDT), intent(inout) :: this
        type(SetupDT), intent(in) :: setup

        call Common_OptionsDT_initialise(this%comm)
        call Optimize_OptionsDT_initialise(this%optimize, setup)

    end subroutine OptionsDT_initialise

    subroutine OptionsDT_copy(this, this_copy)

        implicit none

        type(OptionsDT), intent(in) :: this
        type(OptionsDT), intent(out) :: this_copy

        this_copy = this

    end subroutine OptionsDT_copy

end module mwd_options

File: Pytorch_playground/pycore/derived_type/mwd_options.py
from pycore.derived_type.mwd_cost_options import Cost_OptionsDT
from pycore.derived_type.mwd_optimize_options import Optimize_OptionsDT
from pycore.derived_type.mwd_common_options import Common_OptionsDT
import copy

class OptionsDT:
    def __init__(self, setup, mesh, njoc, njrc):
        self.cost = Cost_OptionsDT(setup, mesh, njoc, njrc)
        self.optimize = Optimize_OptionsDT(setup)
        self.comm = Common_OptionsDT()

    def copy(self):
        return copy.deepcopy(self)